pkg/graph/graph.go

```diff
@@ -46,21 +81,78 @@ func (rm *resourceManager) AddResource(ctx context.Context, resource Resource, b
 		nonNilBlockers = append(nonNilBlockers, blocker)
 
 		_, ok := rm.resources[blocker]
+
 		if !ok {
 			return errors.Errorf("unknown blocker %+v", blocker)
 		}
 	}
 
 	rm.lock.Lock()
-	defer rm.lock.Unlock()
-
-	_, ok := rm.resources[resource]
-	if ok {
+	if _, ok := rm.resources[resource]; ok {
 		return errors.Errorf("resource %+v already added", resource)
 	}
 
-	rm.resources[resource] = nonNilBlockers
-	rm.functions[resource] = run
+	ok, err := rm.preAddResource(ctx, resource)
+	if err != nil {
+		return err
+	}
+
+	if ok {
+		rm.resources[resource] = nonNilBlockers
+		rm.functions[resource] = run
+	}
+	rm.lock.Unlock()
 
 	return nil
 }
+
+func (rm *resourceManager) preAddResource(ctx context.Context, res Resource) (bool, error) {
+	span, _ := opentracing.StartSpanFromContext(ctx, "preAddResource")
+	defer span.Finish()
+
+	var obj client.Object
+	switch resource := res.(type) {
+	case client.Object:
+		obj = resource.DeepCopyObject().(client.Object)
+		gvk := rm.AddGVKToSpan(ctx, span, obj)
+		key := NamespacedNameWithGVK{
+			NamespacedName: types.NamespacedName{
+				Namespace: obj.GetNamespace(),
+				Name:      obj.GetName(),
+			},
+			GroupVersionKind: gvk,
+		}
+
+		if _, ok := rm.primary[key]; ok {
+			return false, nil
+		}
+
+		rm.primary[key] = res
+	case inlineObject:
```

Comment:
This interface only appears here once, we don't need to create an interface type, just use `case interface{GetObject() client.Object}`

Reply:
thx, updated